src/protocol/tcp/src/main/java/org/apache/jmeter/protocol/tcp/proxy/TCPProxyServer.java

```diff
@@ -0,0 +1,110 @@
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.util.Arrays;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TCPProxyServer extends Thread {
+    private static final Logger log = LoggerFactory.getLogger(TCPProxyServer.class);
+
+    ServerSocket serverSocket;
+    public static final int DEFAULT_PORT = 8899;
+    int proxyPort = DEFAULT_PORT;
+    int bufferSize = 4096;
+    boolean runningFlag = true;
+
+    private TCPSamplerManager samplerManager;
+
+    public TCPProxyServer(int proxyPort, TCPSamplerManager samplerManager) {
+        this.proxyPort = proxyPort;
+        this.samplerManager = samplerManager;
+    }
+
+    public void serverStart() throws IOException {
+        log.debug("proxy start");
+        runningFlag = true;
+        serverSocket = new ServerSocket(proxyPort);
+        log.debug("proxy server socket open success");
+        this.samplerManager.managerStart();
+        start();
+        log.debug("proxy start success");
+    }
+
+    public void serverStop() throws IOException {
+        runningFlag = false;
+        serverSocket.close();
+        this.samplerManager.managerStop();
+    }
+
+    @Override
+    public void run() {
+        if (serverSocket == null || serverSocket.isClosed()) {
+            try {
+                serverSocket = new ServerSocket(proxyPort);
+            } catch (IOException ioException) {
+                log.error("start proxy fail " + ioException.getMessage());
+            }
+        }
+        try {
+            while (runningFlag) {
+                Socket socket = serverSocket.accept();
+                log.debug("new client connection " + socket.getInetAddress().getHostName() + ":" + socket.getPort());
+                new Thread(new Runnable() {
+                    @Override
+                    public void run() {
+                        try {
+                            byte[] buffer = new byte[bufferSize];
+                            int readFlag = socket.getInputStream().read(buffer);
+                            if (readFlag != -1) {
+                                byte[] dataRead = Arrays.copyOf(buffer, readFlag);
+                                byte[] responseData = samplerManager.newTCPSampler(new String(dataRead));
+                                socket.getOutputStream().write(responseData);
+                            }
+                        } catch (IOException e) {
+                            log.debug("proxy socket exception " + e.getMessage());
+                        } finally {
+                            try {
+                                socket.close();
+                            } catch (IOException ignore) {
+                            }
+                        }
+                    }
+                }).start();
+            }
+        } catch (SocketException socketException) {
+            log.debug("proxy closed ");
```

Comment:
Add contextual data and exception
log.debug("Proxy closed", socketException)

Reply:
Thank you for your suggestion. I have changed the log style in new commit.